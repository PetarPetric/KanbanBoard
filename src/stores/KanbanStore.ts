import { nextTick, ref } from 'vue';
import { defineStore } from 'pinia';
import type { IColumn } from "../types/Column"
import type { Toast } from "../types/Toast";
import type { ITask } from "../types/Task";
import Repository from "@/services/apiService/APIFactory";
const ColumnRepository = Repository.get("columns");
const TasksRepository = Repository.get("tasks");

export const useKanbanStore = defineStore('KanbanStore', () => {
  const listOfColumns = ref<IColumn[]>([]);
  const toast = ref<null | Toast>(null);
  const columnUpdated = ref(null);

  // Call toast component
  const setToast = (type: string, text: string) => {
    toast.value = {
      type,
      text
    };

    setTimeout(() => {
      toast.value = null;
    }, 1500);
  }

  // Fetch columns
  const fetchColumns = async () => {
    const { data } = await ColumnRepository.get()
    listOfColumns.value = data.sort(
      (a: IColumn, b: IColumn) => a.columnPosition - b.columnPosition
    )
  }

  // Add column api call 
  const addColumn = async (name: string, columnPosition: number) => {
    const form: IColumn = {
      title: name,
      boardId: 0,
      columnPosition,
    }
    await ColumnRepository.create(form);
    await fetchColumns();
    setToast('success', 'Successfully added column');
  }

  // Remove column
  const removeColumn = async (id?: number) => {
    await ColumnRepository.delete(id)
    await fetchColumns();
    setToast('success', 'Successfully removed column');
  }

  // rearrange columns when doing drag and drop like in jira
  const reArrangeColumns = async (id: number, columnPosition: number) => {
    const form = {
      columnPosition
    }
    await ColumnRepository.patch(form, id);
    setToast('info', 'Successfully rearranged columns');
  }

  // Edit columns
  const editColumn = async (id: number, name: string, columnPosition: number) => {
    const form: IColumn = {
      title: name,
      boardId: 0,
      columnPosition,
      id
    }
    await ColumnRepository.update(form, id);
    await fetchColumns();
    setToast('info', 'Edited column');
  }

  // Add task api call
  const addTask = async (columnId: number, title: string, taskPosition: number) => {
    const form: ITask = {
      title,
      description: '',
      columnId,
      taskPosition,
    }
    await TasksRepository.create(form);
    setToast('success', 'Successfully added task');
  }

  // Fetch tasks
  const fetchTasks = async (columnId: number) => {
    const { data } = await TasksRepository.get(columnId)
    return data;
  }

  const fetchSingleTask = async (id: number) => {
    const { data } = await TasksRepository.getSingle(id)
    return data;
  }

  // Move task actions
  const moveTask = async (id: number, columnId?: number, taskPosition?: number) => {
    const form = {
      columnId,
      taskPosition
    }
    const { data } = await TasksRepository.patch(form, id);
    setToast('info', 'Successfully moved task');
    return data;
  }

  // Remove task
  const removeTask = async (id: number) => {
    await TasksRepository.delete(id)
    await fetchColumns();
    setToast('success', 'Successfully removed a task')
  }

  // Edit task
  const editTask = async (form: ITask) => {
    const { data } = await TasksRepository.update(form);
    setToast('info', 'Edited Task');
    return data;
  }

  return {
    fetchColumns,
    addColumn,
    removeColumn,
    editColumn,
    reArrangeColumns,
    addTask,
    fetchTasks,
    fetchSingleTask,
    removeTask,
    moveTask,
    editTask,
    setToast,
    toast,
    listOfColumns,
    columnUpdated
  }
});