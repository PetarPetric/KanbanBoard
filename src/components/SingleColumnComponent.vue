<script setup lang="ts">
import SingleTask from "./SingleTaskComponent.vue";
import { useKanbanStore } from "@/stores/KanbanStore";
import { useDragAndDrop } from "@/composables/dnd";
import { nextTick, computed, ref, watch } from "vue";
import type { IColumn } from "@/types/Column";
import type { ITask } from "@/types/Task";
import {
  BIconXLg,
  BIconPencilSquare,
  BIconCheckLg,
  BIconPlusLg,
} from "bootstrap-icons-vue";
// Store
const store = useKanbanStore();
// Props
const props = defineProps<{
  columnData: IColumn;
  columnIndex: number;
}>();

// DND Composable
const {
  startDrag,
  onDragOver,
  dropElement,
  dragStartItem,
  goingOverItem,
  draggingItemType,
  moveDirection,
  goingOverPosition,
} = useDragAndDrop();

// Column data
const columnTitle = ref("");
const editingColumn = ref(false);
const columnTitleRef = ref<null | HTMLInputElement>(null);
const taskTitleRef = ref<null | HTMLInputElement>(null);

// Tasks data
const addingNewTask = ref(false);
const taskTitle = ref("");
const columnTasks = ref<ITask[]>([]);

// Initialize editing column action
const handleEditColumn = async () => {
  columnTitle.value = props.columnData.title;
  editingColumn.value = true;
  await nextTick();
  if (columnTitleRef.value) {
    columnTitleRef.value.focus();
  }
};

// Handle close button in column actions
const handleCloseBtn = async () => {
  if (editingColumn.value) {
    editingColumn.value = false;
  } else {
    await store.removeColumn(props.columnData?.id);
  }
};

// Edit column action
const editColumn = async () => {
  await store.editColumn(
    props.columnData.id as number,
    columnTitle.value,
    props.columnData.columnPosition
  );
  editingColumn.value = false;
  columnTitle.value = "";
};

// Handle adding task state
const handleAddingTask = async (arg: boolean) => {
  addingNewTask.value = arg;
  await nextTick();
  if (taskTitleRef.value) {
    taskTitleRef.value.focus();
  }
};

// Add task call
const addTask = async () => {
  if (taskTitle.value) {
    let newIndex = columnTasks.value?.length + 1;
    await store.addTask(
      props.columnData.id as number,
      taskTitle.value,
      newIndex
    );
    taskTitle.value = "";
    addingNewTask.value = false;
  }
  fetchTasks();
};

// Fetch Tasks
const fetchTasks = async () => {
  const data = await store.fetchTasks(props.columnData.id as number);
  columnTasks.value = data?.sort(
    (a: ITask, b: ITask) => a.taskPosition - b.taskPosition
  );
};

fetchTasks();
// Number of tasks to show in column
const numberOfTasks = computed(() => {
  if (columnTasks.value && columnTasks.value.length)
    return columnTasks.value.length ? `${columnTasks.value.length} issue` : "";
  else return "";
});

// Animation handling computed method
const handleAnimation = computed(() => {
  if (
    moveDirection.value === "left" &&
    (dragStartItem.value as IColumn)?.columnPosition > 0 &&
    props.columnData.columnPosition <
      (dragStartItem.value as IColumn)?.columnPosition &&
    props.columnData.columnPosition >= goingOverPosition.value
  ) {
    return "all right";
    // if we are moving mouse to right move columns to left
  } else if (
    moveDirection.value === "right" &&
    (dragStartItem.value as IColumn)?.columnPosition <
      store.listOfColumns.length - 1 &&
    props.columnData.columnPosition >
      (dragStartItem.value as IColumn)?.columnPosition &&
    props.columnData.columnPosition <= goingOverPosition.value
  ) {
    return "all left";
  } else return "";
});
</script>

<template>
  <div
    draggable="true"
    @dragend.stop="dropElement(columnData, props.columnIndex)"
    @dragover.stop="
      onDragOver(
        columnData,
        props.columnIndex,
        $event,
        columnTasks,
        columnData.id
      )
    "
    @dragstart.stop="startDrag(columnData, $event)"
    class="column__wrapper"
    data-test="column-wrapper"
    :class="[
      {
        'drag-item':
          columnData.id === dragStartItem?.id &&
          goingOverItem &&
          draggingItemType.column,
        all: draggingItemType.column,
      },
      handleAnimation,
    ]"
  >
    <div class="column__header">
      <input
        ref="columnTitleRef"
        v-model="columnTitle"
        v-if="editingColumn"
        type="text"
      />
      <h5 v-else-if="props?.columnData" class="column__header text-secondary">
        {{ props.columnData.title }} {{ numberOfTasks }}
      </h5>
      <div class="column__header__actions">
        <button
          v-if="editingColumn"
          @click="editColumn"
          data-test="btn-check"
          class="btn"
        >
          <BIconCheckLg />
        </button>
        <button
          v-else
          @click="handleEditColumn"
          data-test="btn-edit"
          class="btn"
        >
          <BIconPencilSquare />
        </button>
        <button @click="handleCloseBtn" data-test="btn-close" class="btn">
          <BIconXLg />
        </button>
      </div>
    </div>
    <div class="column__body">
      <SingleTask
        @removed-task="fetchTasks"
        v-for="(task, i) in columnTasks"
        :single-task="task"
        :task-column="columnTasks"
        :key="i"
        :task-index="i"
      />
      <div v-if="!addingNewTask" class="add-task">
        <button @click="handleAddingTask(true)" class="btn btn--add-task">
          <BIconPlusLg />
          <span class="text-secondary">Create new task</span>
        </button>
      </div>
      <div
        v-else
        :style="`${addingNewTask ? 'display: flex' : ''}`"
        class="actions"
      >
        <input ref="taskTitleRef" v-model="taskTitle" type="text" />
        <button @click="addTask" class="btn btn--check">
          <BIconCheckLg />
        </button>
        <button @click="handleAddingTask(false)" class="btn btn--close">
          <BIconXLg />
        </button>
      </div>
    </div>
  </div>
</template>

<style lang="scss" scoped>
@import "@/styles/singleColumn.scss";
</style>
