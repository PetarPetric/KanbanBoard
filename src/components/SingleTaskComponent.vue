<script setup lang="ts">
import { useKanbanStore } from "@/stores/KanbanStore";
import type { ITask } from "@/types/Task";
import { computed } from "@vue/reactivity";
import { BIconTrash } from "bootstrap-icons-vue";
import { useRouter } from "vue-router";
import { useDragAndDrop } from "@/composables/dnd";
// Emit events
const emit = defineEmits(["removedTask"]);
// Router
const $router = useRouter();
// Store
const store = useKanbanStore();
// Props
const props = defineProps<{
  singleTask: ITask;
  taskIndex: number;
  taskColumn: ITask[];
}>();

const {
  startDrag,
  onDragOver,
  dropElement,
  dragStartItem,
  goingOverItem,
  draggingItemType,
} = useDragAndDrop();

// Get title of column in task
const columnTitle = computed(() => {
  return store.listOfColumns.find(
    (item) => item.id == props.singleTask.columnId
  );
});

// Go To task details
const openTaskDetails = () => {
  $router.push({ path: `/task/${props.singleTask.id}` });
};

// Remov task function
const removeTask = async () => {
  await store.removeTask(props.singleTask.id as number);
  emit("removedTask");
};
</script>

<template>
  <div class="placeholder">
    <div
      draggable="true"
      @dragstart.stop="startDrag(props.singleTask, $event)"
      @dragover.stop="onDragOver(props.singleTask, props.taskIndex, $event)"
      @dragend.stop="
        dropElement(props.singleTask, props.taskIndex, props.taskColumn)
      "
      @click="openTaskDetails"
      data-test="single-task"
      class="single-task"
      :class="[
        {
          'drag-item':
            props.singleTask.id === dragStartItem?.id &&
            goingOverItem &&
            draggingItemType.task,
        },
      ]"
    >
      <div data-test="single-task-header" class="single-task__header">
        <span class="text-secondary">{{ props.singleTask.title }}</span>
        <button
          data-test="btn-remove"
          @click.stop="removeTask"
          class="btn btn--remove"
        >
          <BIconTrash />
        </button>
      </div>
      <div data-test="single-task-body" class="single-task__body">
        <span class="text-secondary">{{ columnTitle?.title }} column</span>
      </div>
    </div>
  </div>
</template>

<style lang="scss" scoped>
@import "@/styles/task.scss";
.placeholder {
  border-radius: 3px;
  margin: 1rem 0;
  height: 80px;
  position: relative;
  background-color: transparent;

  .single-task {
    position: absolute;
  }
}
.down {
  transform: translateY(80px);
}

.up {
  transform: translateY(-80px);
}
</style>
