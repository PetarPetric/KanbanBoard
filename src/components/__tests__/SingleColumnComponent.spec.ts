import { mount, VueWrapper, flushPromises } from "@vue/test-utils";
import { createTestingPinia } from '@pinia/testing'
import { useKanbanStore } from "@/stores/KanbanStore";
import SingleColumnComponent from "@/components/SingleColumnComponent.vue";
import SingleTaskComponent from "@/components/SingleTaskComponent.vue"

describe("single column component", () => {
  let wrapper: VueWrapper<any>;
  let store: any;

  beforeEach(() => {
    wrapper = mount(SingleColumnComponent as any, {
      props: {
        columnData: {
          title: "New Stage",
          boardId: 0,
          columnPosition: 0,
          id: 10
        },
      },
      global: {
        plugins: [createTestingPinia({
          createSpy: vi.fn,
        })],
      }
    });

    store = useKanbanStore();
  });

  it("should render correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render children correctly', async () => {
    // Mock tasks

    wrapper.vm.columnTasks = [{
      title: "First task",
      description: "",
      columnId: 10,
      id: 7,
      taskPosition: 0,
    },
    {
      title: "Second Task",
      description: "",
      columnId: 10,
      id: 8,
      taskPosition: 1,
    },
    {
      title: "Random Task",
      description: "",
      columnId: 10,
      id: 9,
      taskPosition: 2,
    },]

    // wait for dom to update

    await flushPromises();

    expect(wrapper.findAllComponents(SingleTaskComponent).length).toEqual(3)


    expect(wrapper.findAll('[data-test="single-task-header"] span')[0].text()).toContain('First task')
    expect(wrapper.findAll('[data-test="single-task-header"] span')[1].text()).toContain('Second Task')
    expect(wrapper.findAll('[data-test="single-task-header"] span')[2].text()).toContain('Random Task')
  })
});
