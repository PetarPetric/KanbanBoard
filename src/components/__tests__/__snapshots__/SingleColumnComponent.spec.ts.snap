// Vitest Snapshot v1

exports[`single column component > should render correctly 1`] = `
"<div draggable=\\"true\\" class=\\"column__wrapper\\" data-v-973f7a55=\\"\\">
  <div class=\\"column__header\\" data-v-973f7a55=\\"\\">
    <h5 class=\\"column__header text-secondary\\" data-v-973f7a55=\\"\\">New Stage </h5>
    <div class=\\"column__header__actions\\" data-v-973f7a55=\\"\\"><button data-test=\\"btn-edit\\" class=\\"btn\\" data-v-973f7a55=\\"\\"><svg width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 16 16\\" fill=\\"currentColor\\" role=\\"img\\" focusable=\\"false\\" data-v-973f7a55=\\"\\">
          <path d=\\"M15.502 1.94a.5.5 0 0 1 0 .706L14.459 3.69l-2-2L13.502.646a.5.5 0 0 1 .707 0l1.293 1.293zm-1.75 2.456-2-2L4.939 9.21a.5.5 0 0 0-.121.196l-.805 2.414a.25.25 0 0 0 .316.316l2.414-.805a.5.5 0 0 0 .196-.12l6.813-6.814z\\"></path>
          <path fill-rule=\\"evenodd\\" d=\\"M1 13.5A1.5 1.5 0 0 0 2.5 15h11a1.5 1.5 0 0 0 1.5-1.5v-6a.5.5 0 0 0-1 0v6a.5.5 0 0 1-.5.5h-11a.5.5 0 0 1-.5-.5v-11a.5.5 0 0 1 .5-.5H9a.5.5 0 0 0 0-1H2.5A1.5 1.5 0 0 0 1 2.5v11z\\"></path>
        </svg></button><button data-test=\\"btn-close\\" class=\\"btn\\" data-v-973f7a55=\\"\\"><svg width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 16 16\\" fill=\\"currentColor\\" role=\\"img\\" focusable=\\"false\\" data-v-973f7a55=\\"\\">
          <path fill-rule=\\"evenodd\\" d=\\"M13.854 2.146a.5.5 0 0 1 0 .708l-11 11a.5.5 0 0 1-.708-.708l11-11a.5.5 0 0 1 .708 0Z\\"></path>
          <path fill-rule=\\"evenodd\\" d=\\"M2.146 2.146a.5.5 0 0 0 0 .708l11 11a.5.5 0 0 0 .708-.708l-11-11a.5.5 0 0 0-.708 0Z\\"></path>
        </svg></button></div>
  </div>
  <div class=\\"column__body\\" data-v-973f7a55=\\"\\">
    <div class=\\"add-task\\" data-v-973f7a55=\\"\\"><button class=\\"btn btn--add-task\\" data-v-973f7a55=\\"\\"><svg width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 16 16\\" fill=\\"currentColor\\" role=\\"img\\" focusable=\\"false\\" data-v-973f7a55=\\"\\">
          <path fill-rule=\\"evenodd\\" d=\\"M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2Z\\"></path>
        </svg><span class=\\"text-secondary\\" data-v-973f7a55=\\"\\">Create new task</span></button></div>
  </div>
</div>"
`;
